### ISSUE TITLE
Optimizing Cosmology.comoving_distance

### ISSUE THREAD

**Issue by lzkelley** [astropy/astropy#5281]
Lots of cosmological calculations (ones depending on `comoving_distance` at least) are fairly slow on large arrays.  [It looks like this is because the integration is performed repeatedly for every redshift value (I think?)](https://github.com/astropy/astropy/blob/v0.3.x/astropy/cosmology/core.py#L889) --- which is very wasteful.  For example, if I'm looking for the comoving distances at redshifts `z = [0.1, 0.2, 0.5, 1.0]` then the results of all but `max(z)` (in this case `1.0`) are going to be (implicitly) recalculated in the calculation of `max(z)`.   Shouldn't there be a fairly easy way to calculate the results for all input values of `z`, something like:

    # sort redshifts
    inds = np.argsort(z)
    sz = z[inds]
    dc = np.zeros_like(sz)
    integrand = 0.0
    z1 = 0.0
    # Iterate over all redshifts in order
    for ii, z2 in enumerate(sz):
        # Integrate from previous redshift to `z2`
        integrand += integrate_func(z1, z2)
        dc[ii] = integrand_to_com_dist(integrand)

    # *un*sort to match input array
    return dc[np.argsort(inds)]

**Comment by aconley** [comment]
Yes, I've thought a few times that we should be smarter about this.

However -- remember that you can't assume that z is a one dimensional array, so
the above won't work.  There is probably some clever use of numpy functions that
will do it, however.

Having said that, you're probably better off using a ODE solver than a series
of integrals -- and maybe an even better solution would be to support an interpolation
backend for cases like this.

**Comment by sibirrer** [comment]
Is there a plan to implement some of the feature requested in this issue?
I require to compute observables (angular diameter distances) to a set of different redshifts. Currently I do compute the distances separately. This results in the calculation of many very similar coming distance integrals for a significant computational cost. For a given cosmology, there is only one integral required and the partial read out can be cast to predict the observables at different redshifts

I the definition _integral_comoving_distance_z1z2(self, z1, z2) is public I can build my feature on top fully compatible with astropy, otherwise my suggestion would be:
1. in the input for the comoving distance performs a sorting of the redshifts
2. computes a set of partial integrals covering all the differential redshifts that are required
3. add the partial integrals together to compute the array of comoving distances in sorted order
4. re-arrange the output to match the input ordering in redshift

**Comment by aconley** [comment]
Well, somebody needs to find the time and effort to do it.  If you are volunteering, it would certainly be welcome!

As mentioned before, you would likely be better with an ODE solver than just doing a bunch of integrals.  They are better at continuing from some current position more or less by design.

However, before you go that far -- have you considered if interpolation is good enough for your needs?  In most cases that's just fine and is considerably cheaper.  There is at least one example in the cosmology docs, but there are many different ways to set that up.

**Comment by sibirrer** [comment]
@aconley thanks for the prompt and helpful reply! I have an interpolation framework that is now in place. The only (minor) issue I have is that to be still fully consistent with the astropy.cosmology functionality, I need to call private functions with private variables of this class from outside (my interpolation routine class) which can only deal with one-dimensional arrays. It would require quite some design changes to be incorporated in astropy. So I am good for now.
